From ec6b06d7de36ed1f3a1998818aff1c9947a7946b Mon Sep 17 00:00:00 2001
From: lunatickochiya <125438787@qq.com>
Date: Sat, 8 Nov 2025 17:39:57 +0800
Subject: [PATCH] ramips:add nand factory to msg1500

---
 target/linux/ramips/image/mt7621.mk | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/target/linux/ramips/image/mt7621.mk b/target/linux/ramips/image/mt7621.mk
index 7674d0274d..7d5a9b861d 100644
--- a/target/linux/ramips/image/mt7621.mk
+++ b/target/linux/ramips/image/mt7621.mk
@@ -2428,6 +2428,9 @@ define Device/raisecom_msg1500-x-00
   DEVICE_ALT0_VENDOR := Nokia
   DEVICE_ALT0_MODEL := A-040W-Q
   DEVICE_PACKAGES := kmod-mt7615-firmware kmod-usb3 kmod-usb-ledtrig-usbport
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi | \
+	check-size
 endef
 TARGET_DEVICES += raisecom_msg1500-x-00
 
-- 
2.43.0

