From 227e3c9471c94ca7a6f068c8d02d3df3db2205ef Mon Sep 17 00:00:00 2001
From: lunatickochiya <125438787@qq.com>
Date: Sat, 8 Nov 2025 20:37:59 +0800
Subject: [PATCH] kernel: Activate KERNEL_MIPS_FP_SUPPORT for ath79 ramips
 target

---
 config/Config-kernel.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/config/Config-kernel.in b/config/Config-kernel.in
index 64c8c63466..77ebebeb00 100644
--- a/config/Config-kernel.in
+++ b/config/Config-kernel.in
@@ -44,6 +44,8 @@ config KERNEL_DEBUG_FS
 config KERNEL_MIPS_FP_SUPPORT
 	bool
 	default y if TARGET_pistachio
+	default y if TARGET_ramips
+	default y if TARGET_ath79
 
 config KERNEL_ARM_PMU
 	bool
-- 
2.43.0

