diff --git a/net/sstp-client/Makefile b/net/sstp-client/Makefile
index 030081c..95af3cf 100644
--- a/net/sstp-client/Makefile
+++ b/net/sstp-client/Makefile
@@ -45,7 +45,6 @@ Features:
 endef
 
 define Package/sstp-client/conffiles
-/etc/ppp/chap-secrets
 /etc/ppp/peers/peer-sstp-example-nopty.txt
 /etc/ppp/peers/peer-sstp-example.txt
 endef
@@ -58,8 +57,8 @@ define Package/sstp-client/install
 	$(CP) $(PKG_BUILD_DIR)/src/libsstp-api/.libs/*.so* $(1)/usr/lib/
 	$(CP) $(PKG_BUILD_DIR)/src/pppd-plugin/.libs/*.so* $(1)/usr/lib/
 
-	$(INSTALL_DIR) $(1)/etc
-	$(CP) files/etc $(1)
+	$(INSTALL_DIR) $(1)/etc/ppp/peers
+	$(INSTALL_CONF) ./files/etc/ppp/peers/* $(1)/etc/ppp/peers/
 	$(INSTALL_DIR) $(1)/lib/netifd/proto
 	$(INSTALL_BIN) ./files/lib/netifd/proto/sstp.sh $(1)/lib/netifd/proto
 endef
