From 73f7c40b6d33fe601143f81c80dfdfecf0ef15a0 Mon Sep 17 00:00:00 2001
From: lunatickochiya <125438787@qq.com>
Date: Sun, 16 Aug 2026 18:38:21 +0800
Subject: [PATCH 2/7] luci-mod-status:add online users

---
 .../view/status/include/30_network.js         | 32 ++++++++++++++-----
 1 file changed, 24 insertions(+), 8 deletions(-)

diff --git a/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/30_network.js b/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/30_network.js
index 723b73d..f5f86e9 100644
--- a/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/30_network.js
+++ b/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/30_network.js
@@ -14,6 +14,11 @@ const callOdhcp6cStats = rpc.declare({
 	expect: { '': {} },
 });
 
+var callOnlineUsers = rpc.declare({
+        object: 'luci',
+        method: 'getOnlineUsers'
+});
+
 function progressbar(value, max, byte) {
 	const vn = parseInt(value) || 0;
 	const mn = parseInt(max) || 100;
@@ -85,24 +90,35 @@ return baseclass.extend({
 			network.getWANNetworks(),
 			network.getWAN6Networks(),
 			callOdhcp6cStats(),
+			L.resolveDefault(callOnlineUsers(), {})
 		]);
 	},
 
-	render([ct_count, ct_max, wan_nets, wan6_nets, dhcpv6_stats]) {
+	render([ct_count, ct_max, wan_nets, wan6_nets, dhcpv6_stats, onlineusers]) {
 
 		const fields = [
-			{ label: _('Active Connections'), value: ct_max ? ct_count : null }
+			{ label: _('Active Connections'), value: ct_max ? ct_count : null },
+			{ label: _('Online Users'), value: onlineusers ? onlineusers.onlineusers : null }
 		];
 
 		const ctstatus = E('table', { 'class': 'table' });
 
 		for (const { label, value } of fields) {
-			ctstatus.appendChild(E('tr', { 'class': 'tr' }, [
-				E('td', { 'class': 'td left', 'width': '33%' }, [ label ]),
-				E('td', { 'class': 'td left' }, [
-					(value != null) ? progressbar(value, ct_max) : '?'
-				])
-			]));
+			if (label == _('Online Users')) {
+				ctstatus.appendChild(E('tr', { 'class': 'tr' }, [
+					E('td', { 'class': 'td left', 'width': '33%' }, [ label ]),
+					E('td', { 'class': 'td left' }, [
+						(value != null) ? value : '?'
+					])
+				]));
+			} else {
+				ctstatus.appendChild(E('tr', { 'class': 'tr' }, [
+					E('td', { 'class': 'td left', 'width': '33%' }, [ label ]),
+					E('td', { 'class': 'td left' }, [
+						(value != null) ? progressbar(value, ct_max) : '?'
+					])
+				]));
+			}
 		}
 
 		const netstatus = E('div', { 'class': 'network-status-table' });
-- 
2.55.0

