From 783c1e5acae03cf643af83c7f830070a5d985eee Mon Sep 17 00:00:00 2001
From: lunatickochiya <125438787@qq.com>
Date: Sun, 2 Mar 2025 17:05:02 +0800
Subject: [PATCH 3/3] ipt:prefer mesh openssl

---
 include/target.mk | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/target.mk b/include/target.mk
index 02ea68b15c..744dac0a71 100644
--- a/include/target.mk
+++ b/include/target.mk
@@ -50,10 +50,10 @@ DEFAULT_PACKAGES.nas:=\
 # @brief Default packages for @DEVICE_TYPE router.
 ##
 DEFAULT_PACKAGES.router:=\
-	dnsmasq \
-	firewall4 \
-	nftables \
-	kmod-nft-offload \
+	dnsmasq-full \
+	firewall \
+	iptables-zz-legacy \
+	ip6tables-zz-legacy \
 	odhcp6c \
 	odhcpd-ipv6only \
 	ppp \
-- 
2.34.1

