diff --git a/target/linux/ramips/mt7621/target.mk b/target/linux/ramips/mt7621/target.mk
index 2b9a1312af..cf0528efd9 100644
--- a/target/linux/ramips/mt7621/target.mk
+++ b/target/linux/ramips/mt7621/target.mk
@@ -10,7 +10,7 @@ KERNELNAME:=vmlinux vmlinuz
 # make Kernel/CopyImage use $LINUX_DIR/vmlinuz
 IMAGES_DIR:=../../..

-DEFAULT_PACKAGES += wpad-basic-mbedtls uboot-envtools kmod-crypto-hw-eip93
+DEFAULT_PACKAGES += wpad-openssl uboot-envtools kmod-crypto-hw-eip93

 define Target/Description
 	Build firmware images for Ralink MT7621 based boards.
diff --git a/target/linux/ramips/mt7620/target.mk b/target/linux/ramips/mt7620/target.mk
index 34488f95af..1c5cb1af7e 100644
--- a/target/linux/ramips/mt7620/target.mk
+++ b/target/linux/ramips/mt7620/target.mk
@@ -7,7 +7,7 @@ BOARDNAME:=MT7620 based boards
 FEATURES+=usb ramdisk
 CPU_TYPE:=24kc

-DEFAULT_PACKAGES += kmod-rt2800-soc wpad-basic-mbedtls swconfig
+DEFAULT_PACKAGES += kmod-rt2800-soc wpad-openssl swconfig

 define Target/Description
 	Build firmware images for Ralink MT7620 based boards.
--- a/include/target.mk
+++ b/include/target.mk
@@ -52,8 +52,10 @@
 DEFAULT_PACKAGES.router:=\
-	dnsmasq \
-	firewall4 \
-	nftables \
-	kmod-nft-offload \
+	dnsmasq-full \
+	firewall4 \
+	nftables \
+	kmod-nft-offload \
+	iptables-nft \
+	ip6tables-nft \
 	odhcp6c \
 	odhcpd-ipv6only \
 	ppp \
